compiler-cli/src/dependencies.rs

```diff
@@ -1415,3 +1457,272 @@ fn verified_requirements_equality_with_canonicalized_paths() {
+            otp_app: None,
+            requirements: vec!["package_c".into()],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+        ManifestPackage {
+            name: "package_c".into(),
+            version: Version::new(3, 0, 0),
+            build_tools: vec!["gleam".into()],
+            otp_app: None,
+            requirements: vec![],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+        ManifestPackage {
+            name: "package_d".into(),
+            version: Version::new(4, 0, 0),
+            build_tools: vec!["gleam".into()],
+            otp_app: None,
+            requirements: vec!["package_e".into()],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+        ManifestPackage {
+            name: "package_e".into(),
+            version: Version::new(5, 0, 0),
+            build_tools: vec!["gleam".into()],
+            otp_app: None,
+            requirements: vec![],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+    ];
+
+    let packages_to_unlock = vec!["package_a".into(), "package_d".into()];
+    let result = unlock_packages(&packages_to_unlock, &locked, Some(&manifest_packages)).unwrap();
+
+    assert!(!result.contains_key("package_a"));
+    assert!(!result.contains_key("package_b"));
+    assert!(!result.contains_key("package_c"));
+    assert!(!result.contains_key("package_d"));
+    assert!(!result.contains_key("package_e"));
+}
+
+#[test]
+fn test_unlock_packages_empty_input() {
+    let locked = HashMap::from([
+        ("package_a".into(), Version::new(1, 0, 0)),
+        ("package_b".into(), Version::new(2, 0, 0)),
+    ]);
+
+    let manifest_packages = vec![
+        ManifestPackage {
+            name: "package_a".into(),
+            version: Version::new(1, 0, 0),
+            build_tools: vec!["gleam".into()],
+            otp_app: None,
+            requirements: vec!["package_b".into()],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+        ManifestPackage {
+            name: "package_b".into(),
+            version: Version::new(2, 0, 0),
+            build_tools: vec!["gleam".into()],
+            otp_app: None,
+            requirements: vec![],
+            source: ManifestPackageSource::Hex {
+                outer_checksum: Base16Checksum(vec![]),
+            },
+        },
+    ];
+
+    let packages_to_unlock: Vec<EcoString> = vec![];
+    let result = unlock_packages(&packages_to_unlock, &locked, Some(&manifest_packages)).unwrap();
+
+    assert_eq!(result, locked, "Locked packages should remain unchanged when no packages are specified to unlock");
+}
```

Comment:
Could we have a test for packages A, B, and C where A and B both depend on C, and A is being unlocked. B and C should remain locked.

Thank you.

Reply:
My thoughts here was it might actually be desirable to unlock dependency C in this scenario.

By unlocking dependency C, we allow it to update alongside dependency A, while considering the requirements of both dependencies A and B. This should mean C, as part of A's tree, gets an opportunity to update to it's latest allowed version that satisfies all requirements.

I believe that is what is going on at [`compiler-core/src/dependency.rs#L56`](https://github.com/gleam-lang/gleam/blob/main/compiler-core/src/dependency.rs#L56) since dependency C would not be found in `locked` and therefore would be rediscovered via [`compiler-core/src/dependency.rs#L273`](https://github.com/gleam-lang/gleam/blob/main/compiler-core/src/dependency.rs#L273).

Without allowing dependency C to update in this scenario, users would have two choices:

1. Run `gleam deps update dependency_c` (repeating for any downstream dependencies that remain locked)
2. Run `gleam deps update` to update everything.

It's also very possible users may not be aware dependency C exists, if it is a transient and only found within `manifest.toml`, meaning they might not be aware it needs updating.

This led me to believe the proposed changes might be in-line with user's expectations for this command. Perhaps if this is undesirable, we can add an optional flag to direct this behavior? Such as:
```
gleam deps update --no-conflicts <package names>
```

It is possible I've overlooked and/or overthought aspects of this issue. I am happy to add the requested test + changes if you think that is the correct approach.

Reply:
I don't think that `gleam update package1 package2` should impact `package3`.

RE the discovery issue, there's issues for commands to learning about your dependency tree

Reply:
Hmm, one last clarification before I get back to work on this:

Are you wanting `gleam deps update package_a` to only ever unlock/update `package_a`, or is it ok to also unlock/update any of `package_a`'s dependencies so long as no other packages also depend on them?

Reply:
@lpil, I hope you don't mind me following up. I have some time this weekend to work on the changes, but I want to ensure I understand the requirements correctly.

From your last comment, I'm inferring that `gleam deps update <package names>` should never unlock or update packages the user didn't explicitly specify (ie. we would never unlock any package not explicitly named in the command). This would be a significant change from the current implementation of `unlock_packages()` and its associated tests.

To clarify, should we only ever unlock/update the explicitly named packages?

I appreciate your valuable time and guidance on this! Once I have confirmation, I'm ready to implement these changes and update the relevant tests.

Reply:
Sorry, I'm not sure what you mean by this one. You can't tell solely from the list of locked packages like that what is to be unlocked, you need to know what the specified dependencies of the root package are.

Any package that is a dependency of the root package _excluding asked to be upgraded_ should be locked, the rest should be unlocked.

Reply:
I have pushed some changes and additional tests that I think satisfy the requirements. Please review and let me know what you think.